text/0000-offline-buildpackages.md

```diff
@@ -0,0 +1,479 @@
+
+There are two cases depending on weather `<buildpack-uri> points to a buildpackage or a directory containing a `buildpack.toml` file.
+1) Directory containing a `buildpack.toml` file. This method uses the `assets` array to locate download and verify assets.
+2) An existing buildpackage. This will download all assets using `uri`s from the `io.buildpacks.buildpack.layers` label.
+
+The `--config` flag is always optional but recommended to control naming of the generated asset and control which assets are added to the asset-cache. (May also manually pass naming info)
+
+### `asset-cache.toml (optional)`
+
+To provide additional control over asset creation, we may optionally pass  the `pack package-asset-cache` command an additional `--config asset-cache.toml` flag & file, this file defines metadata that impacts the contents of an asset.
+- the `asset-cache.name` field, is used to populate the `io.buildpacks.asset.cache.metadata` field.
+- the `exclude-builpack-assets` array, which must have valid buildpack`id` and `version`, all assets associated with this buildpack will be excluded from the resulting asset cache.
+- `exclude-assets` indicates a `sha256` of a specific asset we should exclude from the resulting asset cache.
+- `asset-override` provides a way to override the `uri` used to download an asset.
+
+
+#### General Format:
+```toml
+[asset-cache (optional)]
+  name = "(required)"
+ 
+[[exclude-buildpack (optional)]]
+  id = "(required)"
+  version = "(required)"
+
+[[exclude (optional)]]
+  sha256 = "(required)"
+
+[[override (optional)]]
+  uri = "preferred/asset/location"
+  sha256 = "asset-sha256"
+```
+
+#### Example
+```toml
+[asset-cache]
+  name = "node-asset-cache"
+ 
+[[exclude-buildpack]]
+  id = "buildpacks/node-engine"
+  version = "1.2.3"
+
+[[exclude]]
+  sha256 = "vulnerable-node10-sha256"
+  name = "Node js version with security vulnerability. (non-functional)"
+
+[[override]]
+  uri = "../local/path/to/nodejs-asset"
+  sha256 = "node13-sha256"
+```
+
+
+Asset cache creation should:
+  - filter excluded assets specified in the `config` file.
+  - Transform all remaining assets to an image layer filesystem changeset where the asset is provided at `/cnb/assets/<artifact-digest>`.
+  - Order all assets layers diffID.
+  - Add the `io.buildpacks.asset.layers` label metadata to the asset image. If no name is specified for `io.buildpacks.asset.cache.metadata`, generate one.
+  - set the created time in image config to a constant value.
+  - set the modification time of all non-asset files to a constant value.
+
+## Metabuildpackage Creation
+
+### Metadata
+
+When creating a metabuildpack, we want to retain all of the asset information from the child buildpacks.
+
+As such `assets` array in `io.buildpacks.buildpack.layers` of the resulting metabuildpack will be a sum of all of the `asset` arrays in its children.
```

Comment:
Why is this neccessary? The `io.buildpacks.buildpack.layers` will already include the children buildpacks as their own entry. It seems that these buildpack entries would be the natural place for the assets. A platform can find the assets for a metabuildpack by looking at each child buildpack's assets.

Reply:
Good point, This is duplicated data.

Reply:
Good point this will be duplicated data, we can remove.